File: accbot-android/app/src/main/java/com/accbot/dca/domain/util/CronUtils.kt
package com.accbot.dca.domain.util

import android.util.Log
import com.cronutils.model.Cron
import com.cronutils.model.CronType
import com.cronutils.model.definition.CronDefinitionBuilder
import com.cronutils.model.time.ExecutionTime
import com.cronutils.parser.CronParser
import java.time.Instant
import java.time.ZoneId
import java.time.ZonedDateTime
import java.util.Locale

/**
 * Utility wrapper around cron-utils for CRON expression parsing,
 * validation, description, and next-execution calculation.
 */
object CronUtils {
    private const val TAG = "CronUtils"

    private val cronDefinition = CronDefinitionBuilder.instanceDefinitionFor(CronType.UNIX)
    private val parser = CronParser(cronDefinition)

    /**
     * Parse a CRON expression. Returns null if invalid.
     */
    private fun parse(expression: String): Cron? {
        return try {
            parser.parse(expression).validate()
        } catch (e: Exception) {
            null
        }
    }

    /**
     * Check if a CRON expression is valid (5-field UNIX CRON).
     */
    fun isValidCron(expression: String): Boolean {
        return parse(expression) != null
    }

    /**
     * Generate a human-readable, localized description of a CRON expression.
     * Handles common DCA patterns with clean output, falls back to CronDescriptor.
     * Returns null if the expression is invalid.
     */
    fun describeCron(expression: String): String? {
        parse(expression) ?: return null
        // Try custom human-readable description first
        describeCronHumanReadable(expression)?.let { return it }
        // Fallback to cron-utils descriptor with current locale
        return try {
            val cron = parse(expression)!!
            val descriptor = com.cronutils.descriptor.CronDescriptor.instance(Locale.getDefault())
            descriptor.describe(cron)
        } catch (e: Exception) {
            Log.w(TAG, "CronDescriptor failed for '$expression', using fallback", e)
            expression
        }
    }

    /**
     * Custom human-readable description for common DCA cron patterns.
     * Returns null for patterns it cannot handle (caller falls back).
     */
    private fun describeCronHumanReadable(expression: String): String? {
        val parts = expression.trim().split("\\s+".toRegex())
        if (parts.size != 5) return null

        val (minuteField, hourField, domField, monthField, dowField) = parts

        // Only handle patterns where month is * (every month)
        if (monthField != "*") return null

        val isCzech = Locale.getDefault().language == "cs"

        // Pattern: */N * * * * → "Every N minutes"
        if (minuteField.startsWith("*/") && hourField == "*" && domField == "*" && dowField == "*") {
            val n = minuteField.removePrefix("*/").toIntOrNull() ?: return null
            return if (isCzech) "Každých $n minut" else "Every $n minutes"
        }

        // Pattern: M */N * * * → "Every N hours"
        if (hourField.startsWith("*/") && domField == "*" && dowField == "*") {
            val n = hourField.removePrefix("*/").toIntOrNull() ?: return null
            return if (isCzech) "Každých $n hodin" else "Every $n hours"
        }

        // Parse specific minutes
        val minutes = parseFieldValues(minuteField) ?: return null
        if (minutes.isEmpty()) return null
        val minute = minutes.first() // use first minute for formatting times

        // Parse specific hours
        val hours = parseFieldValues(hourField) ?: return null
        if (hours.isEmpty()) return null

        // Format time list: "1:30, 13:30"
        val timeStrings = hours.map { h -> "%d:%02d".format(h, minute) }
        val timePart = if (isCzech) {
            timeStrings.joinWithLastSeparator(" a ")
        } else {
            timeStrings.joinWithLastSeparator(" and ")
        }

        // Pattern: M H * * * → "Every day at H:MM"
        if (domField == "*" && dowField == "*") {
            return if (isCzech) "Každý den v $timePart" else "Every day at $timePart"
        }

        // Pattern: M H * * D → "Mon, Wed at H:MM"
        if (domField == "*") {
            val days = parseFieldValues(dowField) ?: return null
            if (days.isEmpty()) return null
            val dayNames = days.map { dayName(it, isCzech) }
            val dayPart = dayNames.joinToString(", ")
            return if (isCzech) "$dayPart v $timePart" else "$dayPart at $timePart"
        }

        // Pattern: M H D * * → "1st, 15th of month at H:MM"
        if (dowField == "*") {
            val doms = parseFieldValues(domField) ?: return null
            if (doms.isEmpty()) return null
            val domPart = if (isCzech) {
                doms.joinToString(", ") { "${it}." }
            } else {
                doms.joinToString(", ") { ordinal(it) }
            }
            return if (isCzech) {
                "$domPart den v měsíci v $timePart"
            } else {
                "$domPart of month at $timePart"
            }
        }

        return null
    }

    /**
     * Parse a cron field into a list of specific integer values.
     * Supports: single values (5), lists (1,13), ranges (1-5).
     * Returns null for *, step patterns, or invalid input.
     */
    private fun parseFieldValues(field: String): List<Int>? {
        if (field == "*" || field.contains("/")) return null
        val values = mutableListOf<Int>()
        for (part in field.split(",")) {
            if (part.contains("-")) {
                val range = part.split("-")
                if (range.size != 2) return null
                val from = range[0].toIntOrNull() ?: return null
                val to = range[1].toIntOrNull() ?: return null
                values.addAll(from..to)
            } else {
                values.add(part.toIntOrNull() ?: return null)
            }
        }
        return values.sorted()
    }

    private fun dayName(dow: Int, czech: Boolean): String {
        // cron: 0=Sun, 1=Mon, ..., 6=Sat  (7=Sun also)
        val normalized = if (dow == 7) 0 else dow
        return if (czech) {
            when (normalized) {
                0 -> "Ne"; 1 -> "Po"; 2 -> "Út"; 3 -> "St"
                4 -> "Čt"; 5 -> "Pá"; 6 -> "So"; else -> "?"
            }
        } else {
            when (normalized) {
                0 -> "Sun"; 1 -> "Mon"; 2 -> "Tue"; 3 -> "Wed"
                4 -> "Thu"; 5 -> "Fri"; 6 -> "Sat"; else -> "?"
            }
        }
    }

    private fun ordinal(n: Int): String = when {
        n % 100 in 11..13 -> "${n}th"
        n % 10 == 1 -> "${n}st"
        n % 10 == 2 -> "${n}nd"
        n % 10 == 3 -> "${n}rd"
        else -> "${n}th"
    }

    private fun List<String>.joinWithLastSeparator(lastSep: String): String = when (size) {
        0 -> ""
        1 -> first()
        else -> dropLast(1).joinToString(", ") + lastSep + last()
    }

    /**
     * Calculate the next execution time after the given instant.
     * Returns null if the expression is invalid.
     */
    fun getNextExecution(expression: String, after: Instant): Instant? {
        val cron = parse(expression) ?: return null
        return try {
            val executionTime = ExecutionTime.forCron(cron)
            val zdt = ZonedDateTime.ofInstant(after, ZoneId.systemDefault())
            executionTime.nextExecution(zdt)
                .orElse(null)
                ?.toInstant()
        } catch (e: Exception) {
            Log.e(TAG, "Failed to calculate next execution for '$expression'", e)
            null
        }
    }

    /**
     * Estimate the average interval in minutes for a CRON expression.
     * Used for monthly cost estimates. Returns null if invalid.
     */
    fun getIntervalMinutesEstimate(expression: String): Long? {
        val cron = parse(expression) ?: return null
        return try {
            val executionTime = ExecutionTime.forCron(cron)
            val now = ZonedDateTime.now()

            // Calculate next 5 executions and average the intervals
            val executions = mutableListOf<ZonedDateTime>()
            var current = now
            repeat(6) {
                val next = executionTime.nextExecution(current).orElse(null) ?: return@repeat
                executions.add(next)
                current = next
            }

            if (executions.size < 2) return 1440L // Default to daily

            val totalMinutes = java.time.Duration.between(executions.first(), executions.last()).toMinutes()
            val intervals = executions.size - 1
            maxOf(totalMinutes / intervals, 1L)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to estimate interval for '$expression'", e)
            1440L // Default to daily
        }
    }
}

File: accbot-ios/AccBot/Domain/Util/CronUtils.swift
import Foundation

/// CRON expression parsing utilities.
/// Supports subset: minute hour day-of-month month day-of-week
enum CronUtils {

    /// Calculate next execution time from a CRON expression.
    /// Uses an optimized algorithm that skips ahead by day/hour when fields don't match,
    /// instead of brute-force minute-by-minute iteration.
    static func getNextExecution(cron: String, from: Date = Date()) -> Date? {
        let parts = cron.trimmingCharacters(in: .whitespaces).split(separator: " ")
        guard parts.count == 5 else { return nil }

        let minuteSpec = String(parts[0])
        let hourSpec = String(parts[1])
        let daySpec = String(parts[2])
        let monthSpec = String(parts[3])
        let weekdaySpec = String(parts[4])

        let calendar = Calendar.current
        var candidate = calendar.date(byAdding: .minute, value: 1, to: from)!

        // Try up to 366 days ahead with smart skipping
        let maxDate = calendar.date(byAdding: .day, value: 366, to: from)!

        while candidate <= maxDate {
            let components = calendar.dateComponents([.minute, .hour, .day, .month, .weekday], from: candidate)
            guard let minute = components.minute,
                  let hour = components.hour,
                  let day = components.day,
                  let month = components.month,
                  let weekday = components.weekday
            else {
                candidate = calendar.date(byAdding: .minute, value: 1, to: candidate)!
                continue
            }

            let cronWeekday = weekday - 1

            // Skip entire month if month doesn't match
            if !matches(value: month, spec: monthSpec) {
                candidate = calendar.date(byAdding: .month, value: 1, to:
                    calendar.date(from: DateComponents(
                        year: components.year, month: month, day: 1, hour: 0, minute: 0
                    ))!
                )!
                continue
            }

            // Skip entire day if day-of-month or day-of-week doesn't match
            if !matches(value: day, spec: daySpec) || !matches(value: cronWeekday, spec: weekdaySpec) {
                candidate = calendar.date(byAdding: .day, value: 1, to:
                    calendar.startOfDay(for: candidate)
                )!
                continue
            }

            // Skip to next hour if hour doesn't match
            if !matches(value: hour, spec: hourSpec) {
                candidate = calendar.date(byAdding: .hour, value: 1, to:
                    calendar.date(from: DateComponents(
                        year: components.year, month: month, day: day, hour: hour, minute: 0
                    ))!
                )!
                continue
            }

            // Check minute
            if matches(value: minute, spec: minuteSpec) {
                return candidate
            }

            candidate = calendar.date(byAdding: .minute, value: 1, to: candidate)!
        }

        return nil
    }

    /// Estimate the average interval in minutes for a CRON expression
    static func getIntervalMinutesEstimate(cron: String) -> Int? {
        let from = Date()
        guard let first = getNextExecution(cron: cron, from: from),
              let second = getNextExecution(cron: cron, from: first)
        else { return nil }
        return Int(second.timeIntervalSince(first) / 60)
    }

    /// Validate a CRON expression
    static func isValid(cron: String) -> Bool {
        let parts = cron.trimmingCharacters(in: .whitespaces).split(separator: " ")
        guard parts.count == 5 else { return false }
        return getNextExecution(cron: cron) != nil
    }

    /// Build a CRON expression from simple components (single hour, optional single day-of-week)
    static func buildCron(minute: Int, hour: Int, dayOfWeek: Int? = nil) -> String {
        let dow = dayOfWeek.map { "\($0)" } ?? "*"
        return "\(minute) \(hour) * * \(dow)"
    }

    /// Build a CRON expression from advanced components (multi-hour, days-of-week/month)
    static func buildCronAdvanced(
        minute: Int,
        hours: Set<Int>,
        daysOfWeek: Set<Int>? = nil,
        daysOfMonth: Set<Int>? = nil
    ) -> String? {
        guard !hours.isEmpty else { return nil }

        let minuteField = "\(minute)"
        let hourField = hours.sorted().map { "\($0)" }.joined(separator: ",")

        if let dows = daysOfWeek, !dows.isEmpty {
            let dowField = dows.sorted().map { "\($0)" }.joined(separator: ",")
            return "\(minuteField) \(hourField) * * \(dowField)"
        }

        if let doms = daysOfMonth, !doms.isEmpty {
            let domField = doms.sorted().map { "\($0)" }.joined(separator: ",")
            return "\(minuteField) \(hourField) \(domField) * *"
        }

        // Daily
        return "\(minuteField) \(hourField) * * *"
    }

    /// Human-readable description of a CRON expression
    static func describeCron(_ expression: String) -> String? {
        let parts = expression.trimmingCharacters(in: .whitespaces).split(separator: " ")
        guard parts.count == 5 else { return nil }

        let minuteField = String(parts[0])
        let hourField = String(parts[1])
        let domField = String(parts[2])
        let monthField = String(parts[3])
        let dowField = String(parts[4])

        guard monthField == "*" else { return nil }

        let isCzech = Locale.current.language.languageCode?.identifier == "cs"

        // Pattern: */N * * * * → "Every N minutes"
        if minuteField.hasPrefix("*/"), hourField == "*", domField == "*", dowField == "*" {
            guard let n = Int(minuteField.dropFirst(2)) else { return nil }
            return isCzech ? "Každých \(n) minut" : "Every \(n) minutes"
        }

        // Pattern: M */N * * * → "Every N hours"
        if hourField.hasPrefix("*/"), domField == "*", dowField == "*" {
            guard let n = Int(hourField.dropFirst(2)) else { return nil }
            return isCzech ? "Každých \(n) hodin" : "Every \(n) hours"
        }

        // Parse specific minutes and hours
        guard let minutes = parseFieldValues(minuteField, range: 0...59), !minutes.isEmpty,
              let hours = parseFieldValues(hourField, range: 0...23), !hours.isEmpty
        else { return nil }

        let minute = minutes.first!
        let timeStrings = hours.sorted().map { String(format: "%d:%02d", $0, minute) }
        let timeList = timeStrings.joined(separator: isCzech ? " a " : " and ")

        let isDomWild = domField == "*"
        let isDowWild = dowField == "*"

        if isDomWild && isDowWild {
            // Daily
            return isCzech ? "Každý den v \(timeList)" : "Every day at \(timeList)"
        }

        if isDomWild && !isDowWild {
            // Days of week
            guard let dows = parseFieldValues(dowField, range: 0...7) else { return nil }
            let normalizedDows = Set(dows.map { $0 == 7 ? 0 : $0 })
            let dayNames = dayOfWeekNames(normalizedDows, czech: isCzech)
            return isCzech ? "\(dayNames) v \(timeList)" : "\(dayNames) at \(timeList)"
        }

        if !isDomWild && isDowWild {
            // Days of month
            guard let doms = parseFieldValues(domField, range: 1...31) else { return nil }
            let dayList = doms.sorted().map { "\($0)." }.joined(separator: ", ")
            return isCzech ? "\(dayList) každého měsíce v \(timeList)" : "\(dayList) of every month at \(timeList)"
        }

        return nil
    }

    // MARK: - Private

    private static func parseFieldValues(_ field: String, range: ClosedRange<Int>) -> [Int]? {
        if field.contains("-") || field.contains("/") || field.contains("*") { return nil }
        let numbers = field.split(separator: ",").map { $0.trimmingCharacters(in: .whitespaces) }
        var result: [Int] = []
        for num in numbers {
            guard let n = Int(num), range.contains(n) else { return nil }
            result.append(n)
        }
        return result.isEmpty ? nil : result
    }

    private static func dayOfWeekNames(_ dows: Set<Int>, czech: Bool) -> String {
        let names: [(Int, String, String)] = [
            (1, "Mon", "Po"), (2, "Tue", "Út"), (3, "Wed", "St"),
            (4, "Thu", "Čt"), (5, "Fri", "Pá"), (6, "Sat", "So"), (0, "Sun", "Ne"),
        ]
        let selected = names.filter { dows.contains($0.0) }
        let joined = selected.map { czech ? $0.2 : $0.1 }.joined(separator: ", ")
        return joined
    }

    private static func matches(value: Int, spec: String) -> Bool {
        if spec == "*" { return true }

        // Handle step (*/n)
        if spec.hasPrefix("*/") {
            guard let step = Int(spec.dropFirst(2)), step > 0 else { return false }
            return value % step == 0
        }

        // Handle range (a-b)
        if spec.contains("-") {
            let parts = spec.split(separator: "-")
            guard parts.count == 2,
                  let low = Int(parts[0]),
                  let high = Int(parts[1])
            else { return false }
            return value >= low && value <= high
        }

        // Handle list (a,b,c)
        if spec.contains(",") {
            let values = spec.split(separator: ",").compactMap { Int($0) }
            return values.contains(value)
        }

        // Exact match
        guard let exact = Int(spec) else { return false }
        return value == exact
    }
}
